[SPARK-28940][SQL] Subquery reuse across all subquery levels

peter-toth opened the pull request:
### What changes were proposed in this pull request?

This PR enables subquery reuse across all subquery levels.
Please note that this feature works for adaptive query execution since https://github.com/apache/spark/pull/25471 but this PR:
- enables reuse for static execution (including Dynamic Partition Pruning) as well
- refactors AQE subquery reuse code a bit

Example query:
```
SELECT (SELECT avg(key) FROM testData), (SELECT (SELECT avg(key) FROM testData))
FROM testData
LIMIT 1
```
Plan before this PR:
```
CollectLimit 1
+- *(1) Project [Subquery scalar-subquery#268, [id=#231] AS scalarsubquery()#276, Subquery scalar-subquery#270, [id=#266] AS scalarsubquery()#277]
   :  :- Subquery scalar-subquery#268, [id=#231]
   :  :  +- *(2) HashAggregate(keys=[], functions=[avg(cast(key#13 as bigint))], output=[avg(key)#272])
   :  :     +- Exchange SinglePartition, true, [id=#227]
   :  :        +- *(1) HashAggregate(keys=[], functions=[partial_avg(cast(key#13 as bigint))], output=[sum#282, count#283L])
   :  :           +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13]
   :  :              +- Scan[obj#12]
   :  +- Subquery scalar-subquery#270, [id=#266]
   :     +- *(1) Project [Subquery scalar-subquery#269, [id=#263] AS scalarsubquery()#275]
   :        :  +- Subquery scalar-subquery#269, [id=#263]
   :        :     +- *(2) HashAggregate(keys=[], functions=[avg(cast(key#13 as bigint))], output=[avg(key)#274])
   :        :        +- Exchange SinglePartition, true, [id=#259]
   :        :           +- *(1) HashAggregate(keys=[], functions=[partial_avg(cast(key#13 as bigint))], output=[sum#286, count#287L])
   :        :              +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13]
   :        :                 +- Scan[obj#12]
   :        +- *(1) Scan OneRowRelation[]
   +- *(1) SerializeFromObject
      +- Scan[obj#12]
```
Plan after this PR:
```
CollectLimit 1
+- *(1) Project [ReusedSubquery Subquery scalar-subquery#241, [id=#148] AS scalarsubquery()#248, Subquery scalar-subquery#242, [id=#164] AS scalarsubquery()#249]
   :  :- ReusedSubquery Subquery scalar-subquery#241, [id=#148]
   :  +- Subquery scalar-subquery#242, [id=#164]
   :     +- *(1) Project [Subquery scalar-subquery#241, [id=#148] AS scalarsubquery()#247]
   :        :  +- Subquery scalar-subquery#241, [id=#148]
   :        :     +- *(2) HashAggregate(keys=[], functions=[avg(cast(key#13 as bigint))], output=[avg(key)#246])
   :        :        +- Exchange SinglePartition, true, [id=#144]
   :        :           +- *(1) HashAggregate(keys=[], functions=[partial_avg(cast(key#13 as bigint))], output=[sum#258, count#259L])
   :        :              +- *(1) SerializeFromObject [knownnotnull(assertnotnull(input[0, org.apache.spark.sql.test.SQLTestData$TestData, true])).key AS key#13]
   :        :                 +- Scan[obj#12]
   :        +- *(1) Scan OneRowRelation[]
   +- *(1) SerializeFromObject
      +- Scan[obj#12]
```

### Why are the changes needed?

Performance improvement.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added new UTs.

peter-toth committed:
[SPARK-28940][SQL] Subquery reuse accross all subquery levels

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15014/
Test PASSed.

peter-toth commented:
cc @maryannxue @cloud-fan

SparkQA commented:
**[Test build #109995 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109995/testReport)** for PR 25644 at commit [`dca5546`](https://github.com/apache/spark/commit/dca5546908096152e396fbc2d1e637b40ecf9f6a).

peter-toth commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/PlanAdaptiveSubqueries.scala:
```diff
@@ -29,8 +29,6 @@ case class PlanAdaptiveSubqueries(
     plan.transformAllExpressions {
       case expressions.ScalarSubquery(_, _, exprId) =>
         subqueryMap(exprId.id)
-      case expressions.InSubquery(_, ListQuery(_, _, exprId, _)) =>
```
@maryannxue I don't know why this case was required so I removed it. Please correct me if I was wrong.

SparkQA commented:
**[Test build #109995 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109995/testReport)** for PR 25644 at commit [`dca5546`](https://github.com/apache/spark/commit/dca5546908096152e396fbc2d1e637b40ecf9f6a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/109995/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15015/
Test PASSed.

SparkQA commented:
**[Test build #109996 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109996/testReport)** for PR 25644 at commit [`fb8d7b8`](https://github.com/apache/spark/commit/fb8d7b86dac9b2f9b285875742ac2d9bfbeb35d7).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15016/
Test PASSed.

AngersZhuuuu commented:
I am doing a plugin for Spark to cache subquery data, hit cached data and replace LogicalPlan, support storage in memory, Alluxio.  some  code do same things. In spark's own  CacheManager,  it don't handle situation like we query hive data, then cache query data, but if cached query's source data has changed, It won't get new data when cache matched.

SparkQA commented:
**[Test build #109996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/109996/testReport)** for PR 25644 at commit [`fb8d7b8`](https://github.com/apache/spark/commit/fb8d7b86dac9b2f9b285875742ac2d9bfbeb35d7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/109996/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15021/
Test PASSed.

peter-toth commented:
> I am doing a plugin for Spark to cache subquery data, hit cached data and replace LogicalPlan, support storage in memory, Alluxio. some code do same things. In spark's own CacheManager, it don't handle situation like we query hive data, then cache query data, but if cached query's source data has changed, It won't get new data when cache matched.

This PR has nothing to do with caching of data, it just enables reuse of subquery results (across all levels) by building up a map of plans. I might use wrong/confusing terminology by calling `subqueryCache` some of my vars. But let's see others' opinion on this.

SparkQA commented:
**[Test build #110001 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110001/testReport)** for PR 25644 at commit [`0f9604c`](https://github.com/apache/spark/commit/0f9604c0662c8503b136496662c954377b3cd60e).

SparkQA commented:
**[Test build #110001 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110001/testReport)** for PR 25644 at commit [`0f9604c`](https://github.com/apache/spark/commit/0f9604c0662c8503b136496662c954377b3cd60e).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110001/
Test FAILed.

peter-toth commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15022/
Test PASSed.

SparkQA commented:
**[Test build #110002 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110002/testReport)** for PR 25644 at commit [`0f9604c`](https://github.com/apache/spark/commit/0f9604c0662c8503b136496662c954377b3cd60e).

SparkQA commented:
**[Test build #110002 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110002/testReport)** for PR 25644 at commit [`0f9604c`](https://github.com/apache/spark/commit/0f9604c0662c8503b136496662c954377b3cd60e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110002/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15186/
Test PASSed.

SparkQA commented:
**[Test build #110177 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110177/testReport)** for PR 25644 at commit [`e7108c5`](https://github.com/apache/spark/commit/e7108c5ba4a12fa3237c7f8ca767c52a70ab41d0).

SparkQA commented:
**[Test build #110177 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110177/testReport)** for PR 25644 at commit [`e7108c5`](https://github.com/apache/spark/commit/e7108c5ba4a12fa3237c7f8ca767c52a70ab41d0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110177/
Test FAILed.

peter-toth committed:
fix dynamic partition pruning test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15193/
Test PASSed.

SparkQA commented:
**[Test build #110187 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110187/testReport)** for PR 25644 at commit [`d96a4eb`](https://github.com/apache/spark/commit/d96a4eb595d2665c4b88b6c786e87e28b1831aab).

maryannxue commented on sql/core/src/test/scala/org/apache/spark/sql/DynamicPartitionPruningSuite.scala:
```diff
@@ -1238,7 +1238,10 @@ class DynamicPartitionPruningSuite
       val plan = df.queryExecution.executedPlan
       val countSubqueryBroadcasts =
-        plan.collectInPlanAndSubqueries({ case _: SubqueryBroadcastExec => 1 }).sum
+        plan.collectInPlanAndSubqueries({
```
Nit: you could also do
```
      val countSubqueryBroadcasts =
        plan.collectInPlanAndSubqueries({ case _: SubqueryBroadcastExec => 1 }).sum
      val countReusedSubqueryBroadcasts =
        plan.collectInPlanAndSubqueries({
          case ReusedSubqueryExec(_: SubqueryBroadcastExec) => 1
        }).sum

      assert(countSubqueryBroadcasts == 1)
      assert(countReusedSubqueryBroadcasts == 1)
```

peter-toth replied:
ok, I changed

maryannxue replied:
We have `InSubqueryExec` now in DPP, so better keep it.

peter-toth replied:
Ok, I've reverted the change.

peter-toth committed:
address review comments, use subquery reuse cache on InSubqueries

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15196/
Test PASSed.

SparkQA commented:
**[Test build #110190 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110190/testReport)** for PR 25644 at commit [`b49d8e1`](https://github.com/apache/spark/commit/b49d8e16f481c165c710124574bed2e374907ee5).

SparkQA commented:
**[Test build #110187 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110187/testReport)** for PR 25644 at commit [`d96a4eb`](https://github.com/apache/spark/commit/d96a4eb595d2665c4b88b6c786e87e28b1831aab).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110187/
Test PASSed.

SparkQA commented:
**[Test build #110190 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110190/testReport)** for PR 25644 at commit [`b49d8e1`](https://github.com/apache/spark/commit/b49d8e16f481c165c710124574bed2e374907ee5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110190/
Test PASSed.

peter-toth committed:
revert InSubquery remove

peter-toth committed:
add InSubqueryExec test

SparkQA commented:
**[Test build #110232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110232/testReport)** for PR 25644 at commit [`336b70a`](https://github.com/apache/spark/commit/336b70a09af5dba0048871641d3719e49ee02146).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15231/
Test PASSed.

SparkQA commented:
**[Test build #110232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110232/testReport)** for PR 25644 at commit [`336b70a`](https://github.com/apache/spark/commit/336b70a09af5dba0048871641d3719e49ee02146).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110232/
Test PASSed.

peter-toth committed:
better DPP subquery reuse, add test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/15242/
Test PASSed.

SparkQA commented:
**[Test build #110245 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110245/testReport)** for PR 25644 at commit [`9800dae`](https://github.com/apache/spark/commit/9800dae0e133924da6fc04305e48869f7dd9e86c).

SparkQA commented:
**[Test build #110245 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110245/testReport)** for PR 25644 at commit [`9800dae`](https://github.com/apache/spark/commit/9800dae0e133924da6fc04305e48869f7dd9e86c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PlanDynamicPruningFilters(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110245/
Test PASSed.

peter-toth commented:
cc @mgaido91, your comments are welcome too

mgaido91 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala:
```diff
@@ -41,9 +41,6 @@ case class InsertAdaptiveSparkPlan(
   private val conf = session.sessionState.conf
 
-  // Subquery-reuse is shared across the entire query.
-  private val subqueryCache = new TrieMap[SparkPlan, BaseSubqueryExec]()
```
we can ermove the import right?

peter-toth replied:
No, unfortunately we can't, `stageCache` still needs it.

mgaido91 reviewed (commented):
the change seems reasonable to me. I don't like too much adding a new map to the QueryExecution but I have no better suggestion at the moment...

peter-toth committed:
Merge commit 'c56a012bc839cd2f92c2be41faea91d1acfba4eb' into SPARK-28940-global-subquery-reuse

# Conflicts:
#	sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15641/
Test PASSed.

peter-toth committed:
Merge commit '8556710409d9f2fbaee9dbf76a2ea70218316693' into SPARK-28940-global-subquery-reuse

peter-toth committed:
fix conflicting ReuseExchange and ReuseSubquery

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16782/
Test PASSed.

SparkQA commented:
**[Test build #111802 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111802/testReport)** for PR 25644 at commit [`280f4c6`](https://github.com/apache/spark/commit/280f4c6014449f807a793821391940d087adc421).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16783/
Test PASSed.

SparkQA commented:
**[Test build #111802 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111802/testReport)** for PR 25644 at commit [`280f4c6`](https://github.com/apache/spark/commit/280f4c6014449f807a793821391940d087adc421).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class PlanDynamicPruningFilters(sparkSession: SparkSession)`
  * `case class PlanSubqueries(sparkSession: SparkSession) extends Rule[SparkPlan] `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111802/
Test PASSed.

peter-toth commented:
After https://github.com/apache/spark/pull/26015 some of the new UTs showed and error and I realized that the new feature of that PR (exchange reuse across all subquery levels) conflicts with my implementation of subquery reuse across all subquery levels.
This was because the `subqueryCache` that I added to the `QueryExecution` contained a cached subquery that was later removed by the new version of `ReuseExchange` rule. The subquery that was cached had a parent `Exchange` node that was replaced to a `ReusedExchangeExec` node.
In the last commit I changed my subquery reuse implementation to something similar to exchange reuse implementation.
I believe this new approach is a bit less efficient but the advantage is that there is no more need for `subqueryCache` in `QueryExecution` and overall the required change is smaller than before.

peter-toth committed:
simplify reuse code

SparkQA commented:
**[Test build #111838 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111838/testReport)** for PR 25644 at commit [`8c77fd9`](https://github.com/apache/spark/commit/8c77fd9af5379e1bef050855ee776fbd66375210).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/16818/
Test PASSed.

SparkQA commented:
**[Test build #111838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/111838/testReport)** for PR 25644 at commit [`8c77fd9`](https://github.com/apache/spark/commit/8c77fd9af5379e1bef050855ee776fbd66375210).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/111838/
Test PASSed.

peter-toth committed:
avoid canonicalization if possible

SparkQA commented:
**[Test build #112175 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112175/testReport)** for PR 25644 at commit [`b84c0d4`](https://github.com/apache/spark/commit/b84c0d40a718a65bbe6187a6b906a958e4f1f1ce).

SparkQA commented:
**[Test build #112179 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112179/testReport)** for PR 25644 at commit [`de6f1a5`](https://github.com/apache/spark/commit/de6f1a5506f62768ee0e946d3788eae4d98f8ae5).

SparkQA commented:
**[Test build #112179 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112179/testReport)** for PR 25644 at commit [`de6f1a5`](https://github.com/apache/spark/commit/de6f1a5506f62768ee0e946d3788eae4d98f8ae5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112179/
Test FAILed.

SparkQA commented:
**[Test build #112175 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112175/testReport)** for PR 25644 at commit [`b84c0d4`](https://github.com/apache/spark/commit/b84c0d40a718a65bbe6187a6b906a958e4f1f1ce).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112175/
Test PASSed.

peter-toth committed:
Merge commit 'd6e33dc3771f30e44f5e9af612b59f011b76fda2' into SPARK-28940-global-subquery-reuse

# Conflicts:
#	sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17727/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112782/
Test FAILed.

peter-toth commented:
retest this please

SparkQA commented:
**[Test build #112844 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112844/testReport)** for PR 25644 at commit [`f86b8c3`](https://github.com/apache/spark/commit/f86b8c3142fba8f4ed9d2e628c4eb085f26e8f8e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/17777/
Test PASSed.

SparkQA commented:
**[Test build #112844 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/112844/testReport)** for PR 25644 at commit [`f86b8c3`](https://github.com/apache/spark/commit/f86b8c3142fba8f4ed9d2e628c4eb085f26e8f8e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/112844/
Test PASSed.
